packages/angular_devkit/architect/src/architect.ts

```diff
@@ -0,0 +1,270 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  BaseException,
+  JsonObject,
+  JsonParseMode,
+  Path,
+  dirname,
+  getSystemPath,
+  join,
+  normalize,
+  parseJson,
+  resolve,
+  schema,
+  virtualFs,
+} from '@angular-devkit/core';
+import { createConsoleLogger } from '@angular-devkit/core/node';
+import { resolve as nodeResolve } from '@angular-devkit/core/node';
+import { Observable } from 'rxjs/Observable';
+import { of } from 'rxjs/observable/of';
+import { _throw } from 'rxjs/observable/throw';
+import { concatMap } from 'rxjs/operators';
+import {
+  BuildEvent,
+  Builder,
+  BuilderClass,
+  BuilderContext,
+  BuilderDescription,
+  BuilderMap,
+} from './builder';
+import { Workspace } from './workspace';
+
+
+export class ProjectNotFoundException extends BaseException {
+  constructor(name?: string) {
+    const nameOrDefault = name ? `Project '${name}'` : `Default project`;
+    super(`${nameOrDefault} could not be found in workspace.`);
+  }
+}
+
+export class TargetNotFoundException extends BaseException {
+  constructor(name?: string) {
+    const nameOrDefault = name ? `Target '${name}'` : `Default target`;
+    super(`${nameOrDefault} could not be found in workspace.`);
+  }
+}
+
+export class ConfigurationNotFoundException extends BaseException {
+  constructor(name: string) {
+    super(`Configuration '${name}' could not be found in project.`);
+  }
+}
+
+export class SchemaValidationException extends BaseException {
+  constructor(errors: string[]) {
+    super(`Schema validation failed with the following errors:\n  ${errors.join('\n  ')}`);
+  }
+}
+
+// TODO: break this exception apart into more granular ones.
+export class BuilderCannotBeResolvedException extends BaseException {
+  constructor(builder: string) {
+    super(`Builder '${builder}' cannot be resolved.`);
+  }
+}
+
+export class WorkspaceNotYetLoadedException extends BaseException {
+  constructor() { super(`Workspace needs to be loaded before Architect is used.`); }
+}
+
+export interface Target<OptionsT = {}> {
+  root: Path;
+  projectType: string;
+  builder: string;
+  options: OptionsT;
+}
+
+export class Architect {
+  private _workspaceSchema = join(normalize(__dirname), 'workspace-schema.json');
+  private _buildersSchema = join(normalize(__dirname), 'builders-schema.json');
```

Comment:
These two seem to never change; should either be readonly, elevated to module level constants, or class static properties (if we want developers to find/use the schemas; e.g., static readonly workspaceSchemaPath = ...).

Reply:
(github got a bit confused and this comment is a duplicate)